### ISSUE TITLE
Native notebooks LaTeX support

### ISSUE THREAD

**Issue by joyceerhl** [microsoft/vscode#119536]
- VS Code Version: 1.55.0-insider
- OS Version: Windows_NT x64 10.0.19042

Steps to Reproduce:

1. Download https://github.com/joyceerhl/notebooks/blob/main/markdown.ipynb. This notebook contains a roundup of LaTeX examples that users have reported did not work in either VS Code notebooks or the Jupyter extension's webview-based notebooks implementation. Each cell contains the LaTeX code and the GitHub issue it was reported in.
2. Open the notebook in the new native notebooks interface in the latest VS Code Insiders with `notebook.experimental.useMarkdownRenderer` set to true.
3. Note that some but not all of the examples are correctly rendered.

Does this issue occur when all extensions are disabled?: Yes

**Comment by mjbvz** [comment]
I don't think we've committed to full latex support. We ship support for katex math formulas to validate the new, extensible markdown renderers. It's still TBD if we would bundle this limited support or require notebook providers to bring their own markdown extensions

**Comment by joyceerhl** [comment]
Following up on Friday's meeting, I did some investigation to root cause the math examples we talked about that aren't rendering with the current native notebooks experimental markdown renderer. Please let me know if I can provide other context that might be more helpful.


## Math in markdown in notebooks

| Notebook offering | Markdown parser | Math library | Additional notes |
| --- | --- | --- | --- |
| JupyterLab | [marked](https://www.npmjs.com/package/marked) | MathJax 2 | JupyterLab preprocesses markdown to hide math expressions from their markdown parser and defer math handling to MathJax. The @jupyterlab/rendermime package provides default renderers for Markdown in Jupyter. [Logic for preprocessing markdown](https://github.com/jupyterlab/jupyterlab/blob/bc55a336f18e964995a5cfc7f3a9c2a71e1388ca/packages/rendermime/src/renderers.ts#L328-L358): remove math, convert markdown to HTML, replace math, then render HTML.  [Logic for preprocessing math](https://github.com/jupyterlab/jupyterlab/blob/master/packages/rendermime/src/latex.ts). JupyterLab also provides extensions for rendering math using KaTeX or MathJax 3: https://github.com/jupyterlab/jupyter-renderers |
| nteract | @nteract/outputs package [vendors](https://github.com/nteract/outputs/blob/master/packages/outputs/src/components/media/markdown.tsx) [@nteract/markdown](https://github.com/nteract/markdown) which uses [react-markdown](https://github.com/remarkjs/react-markdown) | MathJax 2 via [@nteract/mathjax](https://github.com/nteract/mathjax) | @nteract/mathjax is a React wrapper that [pulls MathJax](https://github.com/nteract/mathjax/blob/28479a141d83a1cd2d086f9ac4034540b0d6e4a5/src/provider.tsx#L30) from https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js?config=TeX-MML-AM_CHTML |

(Google Colaboratory notebooks aren't open source but [this comment](https://github.com/googlecolab/colabtools/issues/1008#issuecomment-589765585) indicates they try to be consistent with JupyterLab.)


## Math in markdown in VS Code native notebooks

| Code that doesn't work | Cause | What to do about it |
| --- | --- | --- |
| [Jupyter notebooks documentation on examples of MathJax equations](https://github.com/jupyter/notebook/blob/57fcc413b4cd8e2c15d3fd4d14f79f3f339cfd12/docs/source/examples/Notebook/Typesetting%20Equations.ipynb) | https://katex.org/'s playground correctly renders the source for almost all of these examples, so this is not a syntax parity issue. Current (unvalidated) theory is this works in Jupyter due to the math preprocessing that they perform. | Implement same preprocessing hacks in math renderer extension and see if that suffices |
| [Euler's identity: $$ e^{i \pi} + 1 = 0 $$](https://github.com/microsoft/vscode/issues/119195) | Likely caused by markdown parser. Inline `$$`-delimited math equations are rendered in block style in JupyterLab. Users can work around this by placing the `$$` bound section on a newline. | Implement same preprocessing hacks in math renderer extension and see if that suffices |
| [$ e^{i \pi} + 1 = 0 $](https://github.com/microsoft/vscode/issues/119195) | Known limitation of markdown-it-katex library: doesn't support inline math equations where the equation delimiter `$` is surrounded by spaces. See https://github.com/waylonflinn/markdown-it-katex#syntax | Upstream fix to markdown-it-katex |
| [$$ \begin{eqnarray} \frac{dx}{dt} &=& v, \text{ and} \\\ \frac{dv}{dt} &=& a, \end{eqnarray} $$](https://github.com/microsoft/vscode-jupyter/issues/4231) | Issue provides the same equation in single- and multi-line format. Neither works with experimental markdown renderer support. Single-line variant works in MathJax demo and not in https://katex.org playground. Root cause seems to be a KaTeX-specific limitation: https://github.com/KaTeX/KaTeX/issues/1931. | Add support for MathJax math flavor e.g. using markdown-it-mathjax or [@nteract/mathjax-electron](https://github.com/nteract/mathjax-electron) |
| [$$ \text{For } x \in \mathbb{R}^n \text{, } sigmoid(x) = sigmoid\begin{pmatrix} x_1 \ x_2 \ ... \ x_n \\end{pmatrix} = \begin{pmatrix} \frac{1}{1+e^{-x_1}} \ \frac{1}{1+e^{-x_2}} \ ... \ \frac{1}{1+e^{-x_n}} \\end{pmatrix}\tag{1} $$](https://github.com/microsoft/vscode-jupyter/issues/1434) | LaTeX equation provided doesn't work in [MathJax demo](https://jsbin.com/jafodoqifa/edit?html,output) either, so it doesn't seem to be a syntax parity issue. Replacing `\\end` with `\end` seems to get this particular example working in the MathJax demo and with the experimental markdown renderer in native notebooks. | Possibly invalid LaTeX example |
| https://github.com/microsoft/vscode-jupyter/issues/4938 | Code sample provided doesn't render in JupyterLab, MathJax playground, or KaTeX playground. Including for completeness. | Possibly invalid LaTeX example |
| [LaTeX in HTML in Markdown](https://github.com/microsoft/vscode-jupyter/issues/1328) | Likely requires a fix at the markdown parser level, although I don't know enough about markdown to say for sure. | TBD |



## Other background

Leaving this background info here for anyone else who is starting out with understanding LaTeX in Markdown.
- KaTeX and MathJax are popular libraries for typesetting math on the web. 
- KaTeX is smaller and [up to 100x faster](https://www.intmath.com/blog/mathematics/katex-a-new-way-to-display-math-on-the-web-9445) than MathJax. MathJax is more feature complete. However, neither is fully LaTeX/TeX compliant.
- The KaTeX library was created by Khan Academy because their use case is rendering a lot of simple math equations for users with slower computers: https://aperiodical.com/2014/09/katex-the-fastest-math-typesetting-library-for-the-web/ 
- Lists of things that KaTeX does not support (cobbled together from GitHub issues and possibly out of date): https://github.com/KaTeX/KaTeX/wiki/Things-that-KaTeX-does-not-%28yet%29-support, https://katex.org/docs/support_table.html 
- JupyterLab uses MathJax 2 by default.
- MathJax 3 allegedly mitigates many of the performance problems with MathJax 2.

**Comment by mjbvz** [comment]
@joyceerhl I've forked markdown-it-katex (https://github.com/mjbvz/markdown-it-katex) and applied a few minor fixes

I think the Python team should look into these since Python is the primary consumer of math support. On the VS Code side, we're focusing on extensible markdown renderers right now. Feel free to submit PRs to the upstream repo or let me know if you find a better markdown-math integration library

**Comment by joyceerhl** [comment]
Thanks @mjbvz we agree and we plan to investigate leveraging the notebookMarkdownRenderer contribution point you added, tracked in https://github.com/microsoft/vscode-jupyter/issues/5365

**Comment by mjbvz** [comment]
@joyceerhl Just to make sure you we are on the same page: you shouldn't need to worry about `notebookMarkdownRenderer` all right now. We bundle https://github.com/mjbvz/markdown-it-katex into VS Code, so if your team can submit fixes for the markdown markdown parsing issues you identified, then we'll pull them into VS Code

**Comment by mjbvz** [comment]
A few more findings on KaTeX vs MathJax

## Bundle Size
Tested using our current webpack script (production but no advanced optimization):

#### KaTeX
260kB of JS + 23kB of CSS + lazily loaded fonts. For a simple equation, I see this loading a 30kB font file
 
#### MathJax3
1.7MB all bundled together (using https://github.com/tani/markdown-it-mathjax3)

## Code Loading
Using the above bundles, render times for equations are roughly similar. However MathJax takes significantly longer to evaluate. There were collected inside a VS Code notebook

#### Katex

![Screen Shot 2021-04-07 at 4 30 15 PM](https://user-images.githubusercontent.com/12821956/113948722-5f1b3400-97c2-11eb-9bd9-c3e37c57fb90.png)

#### MathJax

![Screen Shot 2021-04-07 at 4 29 09 PM](https://user-images.githubusercontent.com/12821956/113948735-6e9a7d00-97c2-11eb-9db0-186933967bd3.png)

## Relayout

We've put in a lot of work to avoid having notebook cells shift around while loading or scrolling.

#### KaTeX
With KaTeX, here's a profile showing how equations are rendered:

https://user-images.githubusercontent.com/12821956/113949119-5d05a500-97c3-11eb-9635-4cfd8f0728be.mov

There are essentially three stages:

1. Unrendered formula
2. Mostly correct but not fully styled (unloaded font?)
3. Final render

The only layout shift happens going between step 1 and 2

#### MathJax

Now here's using MathJax3:

https://user-images.githubusercontent.com/12821956/113949232-9fc77d00-97c3-11eb-8b42-13e93dcceca9.mov

Again, there are three stages:

1. Unrendered formula
2. Partially rendered
3. Final render

However in this case there are two layout shifts: one during the first load and another between steps 2 and 3.

The KaTeX rendering is preferable here since it avoids an extra resize

## Comparing LaTeX Support

[Here's a list](https://katex.org/docs/support_table.html) of LaTeX features that KaTeX supports or doesn't support.

[Here's a list](https://docs.mathjax.org/en/latest/input/tex/macros/index.html) of supported LaTeX command for MathJax (I can't find a list of unsupported features and some elements on this list require using built-in MathJax extensions)

Using the developer console, I put together this really poor scraper to generate a list of LaTeX features supported in MathJax but not in Katex:

```js
// Run these two snippets on each page

// https://katex.org/docs/support_table.html
const katexNotSupported = Array.from(document.querySelectorAll('td'))
	.filter(x => { const s = x.querySelector('span'); return s?.textContent === 'Not supported' })
	.map(x => x.previousSibling.textContent)

// https://docs.mathjax.org/en/latest/input/tex/macros/index.html
const mathJaxSupported = new Set(Array.from(document.querySelectorAll('tr td:first-child')).map(x => x.textContent))

// Then on some page with both sets of results
katexNotSupported.filter(x => mathJaxSupported.has(x))
```

This produces the following list:

```
- \abovewithdelims
- \array
- \Arrowvert
- \arrowvert
- \atopwithdelims
- \bbox
- \bracevert
- \buildrel
- \cancelto
- \cases
- \class
- \cssId
- \ddddot
- \dddot
- \DeclareMathOperator
- \definecolor
- \displaylines
- \enclose
- \eqalign
- \eqalignno
- \eqref
- \hfil
- \hfill
- \idotsint
- \iiiint
- \label
- \leftarrowtail
- \leftroot
- \leqalignno
- \lower
- \mathtip
- \matrix
- \mbox
- \mit
- \mmlToken
- \moveleft
- \moveright
- \mspace
- \newenvironment
- \Newextarrow
- \notag
- \oldstyle
- \overparen
- \overwithdelims
- \pmatrix
- \raise
- \ref
- \renewenvironment
- \require
- \root
- \Rule
- \scr
- \shoveleft
- \shoveright
- \sideset
- \skew
- \Space
- \strut
- \style
- \texttip
- \Tiny
- \toggle
- \underparen
- \unicode
- \uproot
```

**Comment by mjbvz** [comment]
I pushed a change that handles a few of the cases listed. Source text:

![Screen Shot 2021-04-08 at 9 30 29 PM](https://user-images.githubusercontent.com/12821956/114128701-c9f76880-98b1-11eb-9357-256c76b38274.png)

Rendered result:

![Screen Shot 2021-04-08 at 9 30 34 PM](https://user-images.githubusercontent.com/12821956/114128711-cf54b300-98b1-11eb-90c4-5be3c29e9cf1.png)


Specifically this addresses:

- `Euler's identity: $$ e^{i \pi} + 1 = 0 $$`
- `$ e^{i \pi} + 1 = 0 $`
- `$$ \text{For } x \in \mathbb{R}^n \text{, } sigmoid(x) = sigmoid\begin{pmatrix} x_1 \ x_2 \ ... \ x_n \end{pmatrix} = \begin{pmatrix} \frac{1}{1+e^{-x_1}} \ \frac{1}{1+e^{-x_2}} \ ... \ \frac{1}{1+e^{-x_n}} \end{pmatrix}\tag{1} $$`


The next easy one I'm looking into is the single `$` math block :

```
$\begin{align}
\dot{x} & = \sigma(y-x) \\
\dot{y} & = \rho x - y - xz \\
\dot{z} & = -\beta z + xy
\end{align}$
```

Please test in the next insiders a let me know if you notice any regressions

/cc @kieferrm @rebornix

**Comment by mjbvz** [comment]
Ok, the majority of examples from the [Jupyter Math example notebook](https://github.com/jupyter/notebook/blob/57fcc413b4cd8e2c15d3fd4d14f79f3f339cfd12/docs/source/examples/Notebook/Typesetting%20Equations.ipynb) now work too:

![Screen Shot 2021-04-08 at 10 07 08 PM](https://user-images.githubusercontent.com/12821956/114131093-c5817e80-98b6-11eb-8b59-1a971c90e701.png)


Not sure if this change made the insiders for tomorrow or not.

The two remaining cases from the Jupyter notebook that don't work due to KaTeX:

- \begin{array}
- \begin{eqnarray}

**Comment by delucca** [comment]
Is the experimental native LaTeX support works in the VSCode version of Github Codespaces?

I'm trying it with the following version:
```
Version: 1.56.0-insider
Commit: 58186680eb97fe979e1d986bcfa983ab0a6228e4
Date: 2021-04-06T10:32:31.104Z
Browser: Mozilla/5.0 (X11; Linux x86_64; rv:87.0) Gecko/20100101 Firefox/87.0
```

Even after enabling `notebook.experimental.useMarkdownRenderer` flag, any LaTeX formula simply won't work (even those reported here that are working).

Any idea why?

**Comment by mjbvz** [comment]
@delucca  Please provide steps to reproduce the issue. Are you sure you are using native notebooks for example?

**Comment by delucca** [comment]
@mjbvz yes I am.

Do you have access to Github Codespaces? If so, you can do the following:

1. Launch a new Codespaces on any repository
2. Switch to insiders version
3. Activate the `notebook.experimental.useMarkdownRenderer` flag
3. Create a new blank Jupyter Notebook with `Ctrl+P`
4. Try to type any formula on a markdown cell

**Comment by mjbvz** [comment]
Oh actually we have disabled the new renderers on web

https://github.com/microsoft/vscode/blob/9afc734800062717d835b9ac63235c0ec8a7bf50/src/vs/workbench/contrib/notebook/browser/notebookEditorWidget.ts

Will revisit once we feel things are looking good on desktop

**Comment by delucca** [comment]
> Oh actually we have disabled the new renderers on web
> 
> https://github.com/microsoft/vscode/blob/9afc734800062717d835b9ac63235c0ec8a7bf50/src/vs/workbench/contrib/notebook/browser/notebookEditorWidget.ts#L353
> 
> Will revisit once we feel things are looking good on desktop

Oh, I see.

There is no way to bypass that setting? I use the Github Codespaces as my primary IDE, and I would like to use this feature

**Comment by mjbvz** [comment]
@joyceerhl https://github.com/microsoft/vscode/commit/cd9a6a48201099f9441d5cd76581197c88ec34f3 Added telemetry to track how often something that looks like unsupported KaTeX directive appears in a native notebook

We'd still like your help analyzing a broader notebook population (i.e. not just those being opened in VS Code) to see which directives are used most commonly so we know what to prioritize

**Comment by bellerofonte** [comment]
> Ok, the majority of examples from the [Jupyter Math example notebook](https://github.com/jupyter/notebook/blob/57fcc413b4cd8e2c15d3fd4d14f79f3f339cfd12/docs/source/examples/Notebook/Typesetting%20Equations.ipynb) now work too:

Hi! I have the VSCode Version: 1.55.2 on MacOS Mojave, Jupyter extension v2021.5.745244803.
Just tried this notebook - **no** equation was rendered properly. Tried to run all cells and then every cell manually.

VSCode output:
<img width="1090" alt="2021-05-03_1621" src="https://user-images.githubusercontent.com/5753506/116882132-bfb85800-ac2c-11eb-80bc-d95beed2d98b.png">

Browser-based notebook worked fine:
<img width="1145" alt="2021-05-03_1622" src="https://user-images.githubusercontent.com/5753506/116882303-f7bf9b00-ac2c-11eb-8ba9-97ac5aace745.png">

What is wrong in my VSCode?

**Comment by mjbvz** [comment]
Closing as we've shipped support for math equations

Please open new issues for any further bugs / feature requests
